[ALICE 3] Add separate class to describe A3 beam pipe + 12-layer A3 layout

mconcas opened the pull request:
In this PR:
1. 12-layer layout for ALICE 3 Silicon Tracker (TRK)
2. Dedicated and simplified description of the double Be beampipe foreseen in new layout as a simple separated _passive_ volume.

Checks and questions are very welcome.
@qgp

mconcas committed:
Add separate class to describe Alice 3 (A3) Beampipe

mconcas committed:
Try to fix codechecker

qgp commented on macro/build_geometry.C:
```diff
@@ -127,16 +128,21 @@ void build_geometry(FairRunSim* run = nullptr)
-    } else {
-      run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe"));
+    if (isActivated("IT3")) {
+      run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe", 1.6f, 0.05f));
```
I think you are missing the `else` branch here to allow for the simulation of the standard setup with `ENABLE_UPGRADES` active.

mconcas replied:
Isn't it the default behaviour?
It enables the changed beampipe only if `-m PIPE ITS3` is required, the Run3 stadnard one otherwise.

mconcas replied:
True, I always manage to mess with this.

qgp replied:
Maybe I am missing something, but I don't think this is correct yet. The standard beam pipe should be used if `ENABLE_UPGRADES` is not set or the standard ITS is used. At the moment it would always be added. I think the code should read something like this:
```suggestion
#ifdef ENABLE_UPGRADES
     if (isActivated("IT3")) {
       run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe", 1.6f, 0.05f));      
     } else { 
       run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe"));
     }
#else
     run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe"));
#endif
```
Does this make sense to you?

mconcas replied:
It does, I updated the PR, thx.

qgp commented on Detectors/Upgrades/ALICE3/TRK/simulation/src/Detector.cxx:
```diff
@@ -130,38 +106,29 @@ void Detector::configITS(Detector* its)
+  layers.emplace_back(std::array<double, 2>{45.f, 132.f});
+  layers.emplace_back(std::array<double, 2>{60.f, 132.f});
+  layers.emplace_back(std::array<double, 2>{80.f, 132.f});
+  layers.emplace_back(std::array<double, 2>{100.f, 132.f});
```
Can we make this information retrievable, e.g. by adding getters to `mLayerRadii`, `mLayerZLen`, ... ? Or do you see any problem with doing that?

mconcas replied:
True, I alsways manage to mess with this.

mconcas replied:
Well, this class will be eventually cleaned up a lot.
Getters are no-issue, point is that this is configured at runtime: either you access them at simulation time (where you have already the information) or it does not make much sense.
When you want to know that at reco/analysis runtime, perhaps, is more consistent to get the quotas from some product of the simulation process, such as the geometry file.

qgp replied:
Not for this PR, but is there an example how this information can be retrieved at analysis time, or more specifically in the tracking macro?

mconcas committed:
Fix default behaviour for -m PIPE

mconcas committed:
Add getters for layers quotas

mconcas committed:
Add override to destructor

mconcas commented:
I realise in terms of reviews this is kind of "no man's land", except for the part that directly affects the plugging into the simulation chain.
Maybe @sawenzel, @shahor02 can comment on that?

mconcas committed:
Fix ENABLE_UPGRADES logic

mconcas committed:
Clang format

sawenzel commented:
For me the PR is fine and can be merged. I usually watch out for approval from someone to take action.

I have one thought for the future though: The configuration is starting to get a bit complex. Maybe we could think about a new feature of "geometry-templates" or baseline configuration templates. These could be standard named collections of modules belonging together. Something like

`o2-sim --template [ Run3 | Run3-barrel-only | ALICE3 | ... ]`

mconcas commented:
> For me the PR is fine and can be merged. I usually watch out for approval from someone to take action.
> 
> I have one thought for the future though: The configuration is starting to get a bit complex. Maybe we could think about a new feature of "geometry-templates" or baseline configuration templates. These could be standard named collections of modules belonging together. Something like
> 
> `o2-sim --template [ Run3 | Run3-barrel-only | ALICE3 | ... ]`

I think this can be a useful shortening for the usual `o2-sim` command.
I also guess in terms of the underlying code it has to be implemented in such a way that some templates should be availalble only if `ENABLE_UPGRADES` is set.

I was also thinking about how to re-linearize the complexity of the `build_geometry.C` macro, which one would prefer it not to be _polluted_ with upgrade-only code. Perhaps it could be possible to make these templates to correspond to some different instantiation of a macro.
Then we set the default to the current one, in order make the change completely transparent.
Just brainstorming.

qgp commented:
> > For me the PR is fine and can be merged. I usually watch out for approval from someone to take action.
> > I have one thought for the future though: The configuration is starting to get a bit complex. Maybe we could think about a new feature of "geometry-templates" or baseline configuration templates. These could be standard named collections of modules belonging together. Something like
> > `o2-sim --template [ Run3 | Run3-barrel-only | ALICE3 | ... ]`
> 
> I think this can be a useful shortening for the usual `o2-sim` command.
> I also guess in terms of the underlying code it has to be implemented in such a way that some templates should be availalble only if `ENABLE_UPGRADES` is set.
I think this is a good idea. But I don't think this should be available only with `ENABLE_UPGRADES` active, e.g. the `Run3-barrel-only` would make sense without anything upgrade related.

> I was also thinking about how to re-linearize the complexity of the `build_geometry.C` macro, which one would prefer it not to be _polluted_ with upgrade-only code. Perhaps it could be possible to make these templates to correspond to some different instantiation of a macro.
There's always a trade-off between having many places with specific configurations and one place with all the options, the former being cleaner if you look in the right place, the latter giving you an overview of everything. For now, the macro still looks reasonable to me, so I don't see an urgent need to add complexity here (for now).
> Then we set the default to the current one, in order make the change completely transparent.
> Just brainstorming.

qgp commented on Detectors/Upgrades/ALICE3/Passive/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+
+o2_add_library(A3DetectorsPassive
```
Are we sure we want to call the targets `A3...` and not `ALICE3`? I would prefer to make sure that people know where this belongs.

mconcas replied:
To me it makes no difference, was trying to shorten a bit the names.

qgp commented on Detectors/Upgrades/ALICE3/Passive/include/A3DetectorsPassive/A3Pipe.h:
```diff
@@ -0,0 +1,67 @@
+{
+namespace passive
+{
+class A3Pipe : public A3PassiveBase
```
Personally, I am not too fond of using prefixes (like A3 here) to group classes. @sawenzel is this common practice in O2? Wouldn't it be cleaner to use a namespace for this purpose?

mconcas replied:
Just to clarify: this is not for grouping classes, it is defining a new class name for the custom beampipe, can be whatever else.

sawenzel replied:
Some extra qualification is definitely needed. I don't mind having it as part of the classname (allows for easy distinction even when we say "using namespace").

qgp replied:
@mconcas Maybe I was not clear, for me ALICE3::Pipe would be cleaner, also for the future. But if @sawenzel says that they actually prefer having prefixes in the class name, I can live with that. Then we should just make sure we use the same prefix everywhere, and at that point I prefer something that you can meaningful grep for, not sure how many unwanted of occurrences of A3 one gets, but your call then ..

mconcas replied:
I'll move to the extended name, good point about "grepping"

sawenzel replied:
I have no strict preference. A namespace is good (but it needs to be alice3 - namespaces are enforced to be lower case).

qgp commented on run/CMakeLists.txt:
```diff
@@ -19,6 +19,7 @@ target_link_libraries(allsim
                                 O2::SimSetup
                                 FairMQ::FairMQ
                                 O2::CPVSimulation
+                                O2::A3DetectorsPassive
```
Duplicating the code here seems risky (divergence)  to me, what about defining a list with the additionally linked libraries when upgrades are enabled and including this in the one call to `target_link_libraries`?

sawenzel replied:
In any case, linking against this needs to be protected for ENABLE_UPGRADES=ON/OFF?

mconcas replied:
It has indeed to be guarded since we don't want any upgrades-related product to be present in Run3 builds.

qgp replied:
Sure, it needs to be guarded. But can't we just add one line to the list instead of duplicating the whole construction (which means all future adjustments need to be done in two places)?
```cmake
$<$<BOOL:${ENABLE_UPGRADES}>:O2::A3DetectorsPassive>
```

mconcas replied:
looks definitely better, I'll try it

mconcas committed:
Improve naming + CMakeLists.txt for upgrades

mconcas committed:
Clang Format

mconcas committed:
Cleanup

mconcas commented:
- Cleaned a bit the redundancy of the Alice 3 specifications in file/libraries naming, keeping it only where it was needed to be diversified by original ones
- Removed the "A3" abbreviation in favour of Alice3. Only initial is uppercase to comply with coding guidelines and ease the lookup of other uppercase macros
- Introduced more compact syntax to link against upgrade-specific libraries and dictionary generation

qgp commented:
@mconcas I don't know if there is anything else I can do, I have approved the changes but it requires an approval from a member of core if I understand correctly.
